src/parse.fs

```diff
@@ -213,15 +212,17 @@ type private ParseImpl() =
                        "point"; "line"; "triangle"; "lineadj"; "triangleadj"
                       ]
                       |> List.map keyword |> choice <?> "Type qualifier"
-        let qualifier = many storage
+    let hlslQualifier = many hlslStorage
+    let specifiedTypeHLSL =
         let generic = ch '<' >>. manyCharsTill anyChar (ch '>')
                    |> opt
                    |>> (function Some s -> "<" + s + ">" | None -> "")
         let typeName = pipe2 (ident |>> (fun id -> id.Name)) generic (+)
         let typeSpec = structSpecifier <|> (typeName |>> Ast.TypeName)
         let arraySizes = many (between (ch '[') (ch ']') expr)
-        pipe4 qualifier typeSpec generic arraySizes (fun tyQ name _ sizes -> Ast.makeType name tyQ sizes)
+        pipe4 hlslQualifier typeSpec generic arraySizes (fun tyQ name _ sizes -> Ast.makeType name tyQ sizes)
 
+    let qualifier = if options.hlsl then hlslQualifier else glslQualifier
```

Comment:
It looks like `qualifier` is a global value that will be evaluated once at the beginning of the program.
You might have to create a parser like `specifiedType` or create a function.